bioconda_utils/docker_utils.py

```diff
@@ -0,0 +1,402 @@
+            conda in a newly-created container. If None, then use the default
+            installed with bioconda_utils.
+
+        build_script_template : str
+            Template that will be filled in with .format(self=self) and that
+            will be run in the container each time build_recipe() is called. If
+            not specified, uses docker_utils.BUILD_SCRIPT_TEMPLATE.
+
+        dockerfile_template : str
+            Template that will be filled in with .format(self=self) and that
+            will be used to build a custom image. Uses
+            docker_utils.DOCKERFILE_TEMPLATE by default.
+
+        use_host_conda_bld : bool
+            If True, then use the host's conda-bld directory. This will export
+            the host's existing conda-bld directory to the docker container,
+            and any recipes successfully built by the container will be added
+            here.
+
+            Otherwise, use `pkg_dir` as a common host directory used across
+            multiple runs of this RecipeBuilder object.
+
+        pkg_dir : str or None
+            Specify where packages should appear on the host. If
+            `use_host_conda_bld` is True, then force `pkg_dir` to use the
+            host's conda-bld directory.
+
+            Otherwise, if `pkg_dir` is None, then a temporary directory will be
+            used for each call to `RecipeBuilder.build()`. If `pkg_dir` is
+            a string, then it will be created if needed and this directory will
+            be used store all built packages on the host.
+
+            In all cases, `pkg_dir` will be mounted to `container_staging` in
+            the container.
+        """
+        self.image = image
+        self.tag = tag
+        self.requirements = requirements
+        self.conda_build_args = ""
+        self.build_script_template = build_script_template
+        self.dockerfile_template = dockerfile_template
+        uid = os.getuid()
+        usr = pwd.getpwuid(uid)
+        self.user_info = dict(
+            uid=uid,
+            gid=usr.pw_gid,
+            groupname=grp.getgrgid(usr.pw_gid).gr_name,
+            username=usr.pw_name)
+        self.container_recipe = container_recipe
+        self.container_staging = container_staging
+
+        self.host_conda_bld = get_host_conda_bld()
+
+        if use_host_conda_bld:
+            self.pkg_dir = self.host_conda_bld
+        else:
+            if pkg_dir is None:
+                self.pkg_dir = tempfile.mkdtemp()
+            else:
+                if not os.path.exists(pkg_dir):
+                    os.makedirs(pkg_dir)
+                self.pkg_dir = pkg_dir
+
+        self._pull_image()
+        self._build_image()
+
+    def _pull_image(self):
+        """
+        Separate out the pull step to provide additional logging info
+        """
+        logger.info('Pulling docker image %s', self.image)
+        p = sp.run(['docker', 'pull', self.image], stdout=sp.PIPE,
```

Comment:
Just out of curiosity, because I am considering this for Snakemake: Did you experience problems with docker-py, or why did you decide to use plain system calls instead?